### PR TITLE
rewrite coap gateway

### PR DESCRIPTION
**Pull request by LenaLenaPan**

### CODE REVIEW

File: en_US/gateway/coap.md [hunk 1]
```diff
@@ -134,67 +203,25 @@ curl -X 'POST' \
...
-
-
-## Client libraries
-
-- [libcoap](https://github.com/obgm/libcoap)
-- [californium](https://github.com/eclipse/californium)
-
-
-## Appendix: Client Side Interfaces
+## Reference: CoAP Client Guide in Connection Mode
```
Comment (LenaLenaPan) [inline en_US/gateway/coap.md]:
```suggestion
## Reference: CoAP Client Guide
```

File: en_US/gateway/coap.md [hunk 2]
```diff
@@ -2,20 +2,56 @@
...
+| Authentication    | ×                   | √               |
 
-It can also be enabled via the HTTP API, and emqx.conf e.g:
+<!--a brief introduction of the architecture-->
 
-:::: tabs type:card
+## Enable the CoAP Gateway
 
-::: tab HTTP API
+In EMQX 5.0, CoAP gateways can be configured and enabled through the Dashboard, HTTP API, and configuration file `emqx.conf`. This section takes the configuration via Dashboard as an example to illustrate the operating steps.
```
Comment (hjianbo) [inline en_US/gateway/coap.md]: typo?
CoAP gateways -> CoAP gateway

File: en_US/gateway/coap.md [hunk 3]
```diff
@@ -37,70 +73,99 @@ curl -X 'PUT' 'http://127.0.0.1:18083/api/v5/gateway/coap' \
...
+- **Notification Message Type**: Set the type of CoAP messages to be delivered, default: **qos**; optional values: 
 
-```properties
-# emqx.conf
-gateway.coap {
+  - **qos**: Whether the CoAP notification needs to be acknowledged depends on the QoS level of the received message, 
+    - QoS 0, no acknowledgment is required from the client, 
+    - QoS 1/2, acknowledgment is required from the client. 
+  - **con**: The CoAP notification should be acknowledged by the client. 
+  - **non**: The CoAP Notification need not be acknowledged by the client.
```
Comment (hjianbo) [inline en_US/gateway/coap.md]:
```suggestion
  - **non**: The CoAP notification need not be acknowledged by the client. 
```

File: en_US/gateway/coap.md [hunk 4]
```diff
@@ -37,70 +73,99 @@ curl -X 'PUT' 'http://127.0.0.1:18083/api/v5/gateway/coap' \
...
-## Authentication
+  - **coap**: Follow the setting in **Notification Message Type** to determine the QoS level
+    - Use QoS 0 if no acknowledgment is required.
+    - Use QoS 1 if acknowledgment is required.
+  - **qos0**, **qos1**, **qos2**
 
-Only available in `Connection Mode`.
+- **Publish QoS**: Set the default QoS level for publish requests, default: **coap**, optional values: **coap**, **qos0**, **qos1**, **qos2**; the values here 
+
+- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different listeners, for example, *mqttsn*.
```
Comment (hjianbo) [inline en_US/gateway/coap.md]:
```suggestion
- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different protocols, for example, *MQTT-SN*.
```

File: en_US/gateway/coap.md [hunk 5]
```diff
@@ -37,70 +73,99 @@ curl -X 'PUT' 'http://127.0.0.1:18083/api/v5/gateway/coap' \
...
+
+  **Note**: The prefixed string will be removed from the topic name when the message is delivered to the subscriber.
```
Comment (hjianbo) [inline en_US/gateway/coap.md]: It is more accurate, I think:

This topic prefix is managed by the gateway. CoAP clients do not need to add this prefix explicitly when publishing and subscribing.

File: en_US/gateway/coap.md [hunk 6]
```diff
@@ -37,70 +73,99 @@ curl -X 'PUT' 'http://127.0.0.1:18083/api/v5/gateway/coap' \
...
-Only available in `Connection Mode`.
+- **Publish QoS**: Set the default QoS level for publish requests, default: **coap**, optional values: **coap**, **qos0**, **qos1**, **qos2**; the values here 
+
+- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different listeners, for example, *mqttsn*.
+
+  **Note**: The prefixed string will be removed from the topic name when the message is delivered to the subscriber. 
+
+### Add Listeners 
+
+By default, one UPD listener with the name of **default** is already configured on port `5683`, which supports up to 1,024,000 concurrent connections. You can click **Settings** for more customized settings for **Delete** to delete the listener. Or click **Add Listener** to add a new listener.
```
Comment (hjianbo) [inline en_US/gateway/coap.md]: Typos:
UPD listener -> UDP listener

File: en_US/gateway/coap.md [hunk 7]
```diff
@@ -37,70 +73,99 @@ curl -X 'PUT' 'http://127.0.0.1:18083/api/v5/gateway/coap' \
...
+
+<img src="/Users/lena/Documents/GitHub/rebalancing/emqx-docs/en_US/gateway/assets/mqttsn-listerner.png" alt="MQTTSN listener" style="zoom:50%;" />
```
Comment (hjianbo) [inline en_US/gateway/coap.md]: Invalid image path

**Review by hjianbo — CHANGES_REQUESTED** [review]

File: en_US/gateway/coap.md (hunk 2 above)
Comment (LenaLenaPan) [inline en_US/gateway/coap.md]:
```suggestion
In EMQX 5.0, CoAP gateway can be configured and enabled through the Dashboard, HTTP API, and configuration file `emqx.conf`. This section takes the configuration via Dashboard as an example to illustrate the operating steps. 
```

File: en_US/gateway/coap.md (hunk 5 above)
Comment (LenaLenaPan) [inline en_US/gateway/coap.md]:
```suggestion
  **Note**: This topic prefix is managed by the gateway. CoAP clients do not need to add this prefix explicitly when publishing and subscribing.
```

File: en_US/gateway/coap.md (hunk 6 above)
Comment (LenaLenaPan) [inline en_US/gateway/coap.md]:
```suggestion
By default, one UDP listener with the name of **default** is already configured on port `5683`, which supports up to 1,024,000 concurrent connections. You can click **Settings** for more customized settings for **Delete** to delete the listener. Or click **Add Listener** to add a new listener.
```

File: en_US/gateway/coap.md (hunk 7 above)
Comment (LenaLenaPan) [inline en_US/gateway/coap.md]:
```suggestion
<img src="./assets/mqttsn-listerner.png" alt="MQTTSN listener" style="zoom:50%;" />
```

File: en_US/gateway/mqttsn.md [hunk 8]
```diff
@@ -109,13 +109,13 @@ In the **Basic Configuration** tab, you can customize your gateway ID, predefine
   - **Topic ID**: Set the topic ID, which should be an integer between 1 and 65535.
   - **Topic**: Set the topic names.<!--, multiple topics can be added here, separated with a `,`-->
 
-- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different listeners, for example, *mqttsn*.
+- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different protocols, for example, *MQTT-SN*.
```
Comment (hjianbo) [inline en_US/gateway/mqttsn.md]:
```suggestion
- **MountPoint**: Set a string that is prefixed to all topics when publishing or subscribing, providing a way to implement message routing isolation between different protocols, for example, `mqttsn/`.
```

File: en_US/gateway/mqttsn.md [hunk 9]
```diff
@@ -109,13 +109,13 @@ In the **Basic Configuration** tab, you can customize your gateway ID, predefine
...
 
-  **Note**: The prefixed string will be removed from the topic name when the message is delivered to the subscriber. 
+  **Note**: This topic prefix is managed by the gateway. CoAP clients do not need to add this prefix explicitly when publishing and subscribing.
```
Comment (hjianbo) [inline en_US/gateway/mqttsn.md]:
```suggestion
  **Note**: This topic prefix is managed by the gateway. MQTT-SN clients do not need to add this prefix explicitly when publishing and subscribing.
```

**Review by hjianbo — APPROVED** [review]

**Review by wivwiv — APPROVED** [review]

**Pull request merged**